### PR TITLE
move the nullifying error messages to extensions

### LINKED ISSUE #2071: Move the "Cannot return null for non-nullable field" error to extensions
this error message was added to help in debugging what happened in query execution, because it happened too often that a large part of the response would be nullified, without any information about the cause. This has created confusion though, because some clients treat it as a hard error, while nullification rules are part of normal operations. Moreover, when a subgraph returns a response with a non null field that's absent or nullified, there's usually an error with the message.
We should move the `Cannot return null for non-nullable field` message to extensions instead, to give more context on query execution but in a quiet way.

### PR DESCRIPTION
**Pull request by Geal**
Fix #2071

Adding them in the list of errors was potentially redundant (subgraph can already add an error message indicating why a field is null) and could be treated as a failure by clients, while nullifying fields is a part of normal operation. We still add the messages in extensions so clients can easily debug why parts of the response were removed

### CODE REVIEW

**Comment by Geal** [comment]
now accepting naming suggestions for the new extensions field. I put "formatting", but maybe that could be "nullability"? WDYT?

**Comment by bnjjj** [comment]
I prefer `nullability` over `formatting` because it explains what's the name of the rules. And if you search for `nullability` in the specs you have some better results.

File: apollo-router/src/spec/query.rs [hunk 1]
```diff
@@ -81,7 +81,12 @@ impl Query {
                                 },
                             );
 
-                            response.errors.extend(parameters.errors.into_iter());
+                            if !parameters.errors.is_empty() {
+                                response.extensions.insert(
+                                    "formatting",
```
Comment (garypen) [inline apollo-router/src/spec/query.rs:86]: Is "formatting" the correct name? Would "composition" be a better choice or "assembly". Is there an established convention for this kind of naming?
UPDATE: I just saw the other comments about naming. I don't think "nullability" is a good name, but could be convinced.

File: apollo-router/src/services/supergraph_service.rs [hunk 2]
```diff
@@ -533,6 +533,41 @@ mod tests {
...
+            .unwrap()
+            .schema(SCHEMA)
+            .extra_plugin(subgraphs)
+            .build()
+            .await
+            .unwrap();
+
+        let request = supergraph::Request::fake_builder()
+            .query("query { currentUser { activeOrganization { id creatorUser { name } } } }")
+            // Request building here
```
Comment (garypen) [inline apollo-router/src/services/supergraph_service.rs:557]: You can probably remove this comment.

File: apollo-router/src/services/supergraph_service.rs [hunk 3]
```diff
@@ -492,7 +492,7 @@ mod tests {
    @join__owner(graph: ORGA)
    @join__type(graph: ORGA, key: "id")
    @join__type(graph: USER, key: "id") {
-       id: ID
+       id: ID!
```
Comment (garypen) [inline apollo-router/src/services/supergraph_service.rs:495]: Hmm, that seems like a material change which isn't related to this PR. Would you update the PR text to indicate why this is changing?

**Review by garypen — COMMENTED** [review]
a few questions/comments about the general design. The code looks good.

**Review by BrynCooke — APPROVED** [review]

File: apollo-router/src/services/supergraph_service.rs (hunk 3 above)
Comment (Geal) [reply apollo-router/src/services/supergraph_service.rs:501]: this is only in a test, to have a field that can trigger nullability rules

File: apollo-router/src/spec/query.rs (hunk 1 above)
Comment (BrynCooke) [reply apollo-router/src/spec/query.rs:140]: Possibly `value-completion`? As this is the part of the spec that handles nullability https://spec.graphql.org/June2018/#sec-Value-Completion

Comment (o0Ignition0o) [reply apollo-router/src/spec/query.rs]: cc @Geal before you squashmerge ^^

Comment (Geal) [reply apollo-router/src/spec/query.rs]: that works for me

**Review by garypen — APPROVED** [review]

**Pull request merged**